athena-storage-api/src/main/java/com/amazonaws/athena/storage/datasource/ParquetDatasource.java

```diff
@@ -0,0 +1,430 @@
+            MessageType messageType = reader.getFileMetaData().getSchema();
+            Configuration configuration = new Configuration();
+            configuration.set(ReadSupport.PARQUET_READ_SCHEMA, messageType.toString());
+            final MessageColumnIO columnIO = new ColumnIOFactory().getColumnIO(messageType);
+            List<GroupSplit> groupSplits = storageSplit.getGroupSplits();
+            for (GroupSplit groupSplit : groupSplits) {
+                PageReadStore pages = reader.readRowGroup(groupSplit.getGroupIndex());
+                FilterCompat.Filter filter = FilterCompat.get(page(groupSplit.getRowOffset(), groupSplit.getRowCount()));
+                if (pages != null && queryStatusChecker.isQueryRunning()) {
+                    ParquetFilter parquetFilter = new ParquetFilter(schema, messageType, split);
+                    ConstraintEvaluator evaluator = parquetFilter.evaluator(tableInfo, split, constraints);
+                    LOGGER.debug("Parquet evaluator: {}", evaluator);
+                    storgeGroupRecordConverter = new StorgeGroupRecordConverter(messageType, evaluator);
+                    addRecords(schema, messageType, columnIO.getRecordReader(pages,
+                            storgeGroupRecordConverter, filter), objectName, spiller, queryStatusChecker);
+                }
+            }
+        }
+        finally {
+            timer.stop();
+            LOGGER.debug("Time took to read records: {}", timer.elapsed(TimeUnit.SECONDS));
+        }
+    }
+
+    /**
+     * Add a record when not being filtered
+     *
+     * @param messageType       An instance of Schema with selected fields and associates types
+     * @param groupRecordReader A reader to field/value from Group
+     */
+    private void addRecords(Schema schema, MessageType messageType, RecordReader<Group> groupRecordReader, String partFileName,
+                            BlockSpiller spiller, QueryStatusChecker queryStatusChecker)
+    {
+        Group group;
+            TypeFactory.ValueResolver valueResolver = TypeFactory.valueResolver(messageType);
+        while ((group = groupRecordReader.read()) != null) {
+            if (queryStatusChecker.isQueryRunning()
+                    && (storgeGroupRecordConverter == null || !storgeGroupRecordConverter.shouldSkipCurrent())) {
+                LOGGER.debug("Parquet record group: {}, class name: {}", group, group.getClass().getName());
+                Map<String, Object> record = valueResolver.getRecord(group);
+                record.put(BLOCK_PARTITION_COLUMN_NAME, partFileName);
+                spiller.writeRows((Block block, int rowNum) -> {
+                    boolean isMatched = true;
+                    for (final Field field : schema.getFields()) {
+                        Object fieldValue = record.get(field.getName());
+                        isMatched &= block.offerValue(field.getName(), rowNum, fieldValue);
+                    }
```

Comment:
Why are we writing into Arrow row by row? Parquet and Arrow are columnar data formats, and both contain native libraries that make it extremely efficient to convert between the two. Writing row-by-row quite literally removes the benefit of a columnar format.

Reply:
Ok, so I figured out that Arrow 10.0.0 has S3 support built into the `arrow-dataset` library but does not have GCS enabled.

And using the s3 support, we can actually hit the GCS endpoint by using endpoint override: https://arrow.apache.org/docs/r/articles/fs.html#uri-options

They explain and demonstrate the GCS S3 compatible endpoint that will act as if its S3 to any client (including Arrow).

The only thing is that we would require users to generate and store a Google HMAC key in AWS Secrets Manager to access their GCS bucket through this endpoint.

Link about Google HMAC keys: https://cloud.google.com/storage/docs/authentication/hmackeys

The URI used would then look something like this:

```
s3://gcs_hmac_access_key:gcs_hmac_secret_key@some_gcs_bucket/some_gcs_object.parquet?endpoint_override=https%3A%2F%2Fstorage.googleapis.com&allow_bucket_creation=false
```

Reply:
Here is some sample code from my test setup:

```
import org.apache.arrow.dataset.file.FileFormat;
import org.apache.arrow.dataset.file.FileSystemDatasetFactory;
import org.apache.arrow.dataset.jni.NativeMemoryPool;
import org.apache.arrow.dataset.scanner.ScanOptions;
import org.apache.arrow.dataset.scanner.Scanner;
import org.apache.arrow.dataset.source.Dataset;
import org.apache.arrow.dataset.source.DatasetFactory;
import org.apache.arrow.memory.BufferAllocator;
import org.apache.arrow.memory.RootAllocator;
import java.util.stream.StreamSupport;
import org.apache.arrow.vector.VectorSchemaRoot;
import org.apache.arrow.vector.ipc.ArrowReader;


String uri = "s3://some-example-bucket/fulfillment/year=2022/month=7/day=1/hour=11/part-00001-870c6afb-3ff1-488c-8043-6f0b50c22a01.c000.snappy.parquet";
ScanOptions options = new ScanOptions(/*batchSize*/ 32768);

try (
    BufferAllocator allocator = new RootAllocator();
    DatasetFactory datasetFactory = new FileSystemDatasetFactory(allocator, NativeMemoryPool.getDefault(), FileFormat.PARQUET, uri);
    Dataset dataset = datasetFactory.finish();
    Scanner scanner = dataset.newScan(options);
    ArrowReader reader = scanner.scanBatches()
) {
    //Schema schema = datasetFactory.inspect();
    //System.out.println(schema);
    while (reader.loadNextBatch()) {
      try (VectorSchemaRoot root = reader.getVectorSchemaRoot()) {
        System.out.print(root.contentToTSVString());
      }
    }
    System.out.println("_______________________________________________________________________*********");
} catch (Exception e) {
    e.printStackTrace();
}
```

Output from that sample code:
![s3_parquet_output](https://user-images.githubusercontent.com/136857/202020194-4c66de40-5462-4516-8879-bf1a7bd2e1f6.png)

Reply:
I wanted to also note that you should try to make use of projections as well when possible: https://arrow.apache.org/docs/java/dataset.html#projection